<a href="https://colab.research.google.com/github/jorozcov/IO1/blob/main/IO1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problema**

**Minimizar 3*x + 8*y**

sujeto a:

x + y >= 50

x - y <= 20

x >= 0
y >= 0

In [4]:
from cvxpy.reductions.solvers import solver
import cvxpy as cp

# Crear variables
x = cp.Variable()
y = cp.Variable()

# Restricciones

restricciones = [x + y >= 50,
                 x - y <= 20,
                 x >= 0,
                 y >= 0]

# Función objetivo
FO = cp.Minimize(3*x + 8*y)

#Plantear problema y resolverlo
prob = cp.Problem(FO, restricciones)


prob.solve(solver= cp.OSQP) 

print("Optimización con OSQP")
print("Estado :", prob.status)
print("Valor óptimo", prob.value)
print("X óptima =", x.value)
print("Y óptima =", y.value)
print()


prob.solve(solver= cp.CVXOPT) 

print("Optimización con CVXOPT")
print("Estado :", prob.status)
print("Valor óptimo", prob.value)
print("X óptima =", x.value)
print("Y óptima =", y.value)

Optimización con OSQP
Estado : optimal
Valor óptimo 225.0
X óptima = 35.0
Y óptima = 15.0

Optimización con CVXOPT
Estado : optimal
Valor óptimo 224.99999976401097
X óptima = 35.00000000026747
Y óptima = 14.999999970401069


# **Comparación de los resultados**

Practicamente con ambos solver se obtuvo el mismo resultado, si se redondea, el valor optimo es 225 con x = 35 y y = 15, no hay diferencia significativa para decidir cuál es el óptimo. Ambos solver están hechos para la solución de problemas convexos, es decir, problemas de optimización continuos; y este, al ser un problema de programación lineal, es perfectamente viable tanto con OSQP o con CVXOPT. 




Si se mira desde un lado extremo, la solución con el solver OSQP es "mejor", ya que converge a la solución exacta sin necesidad de redondeos, pero a los ojos de un analista ambos solver dan la misma solución.